src/pybind/mgr/crash/module.py

```diff
@@ -50,65 +128,115 @@ def timestamp_filter(self, f):
         :returns: crash reports for which f(time) returns true
         """
         def inner(pair):
-            _, meta = pair
-            meta = json.loads(meta)
-            time = self.time_from_string(meta["timestamp"])
+            _, crash = pair
+            time = self.time_from_string(crash["timestamp"])
             return f(time)
-        matches = filter(inner, six.iteritems(
-            self.get_store_prefix("crash/")))
-        return [(k, json.loads(m)) for k, m in matches]
+        return filter(inner, self.crashes.items())
 
     # command handlers
 
     def do_info(self, cmd, inbuf):
         crashid = cmd['id']
-        key = 'crash/%s' % crashid
-        val = self.get_store(key)
-        if not val:
+        crash = self.crashes.get(crashid)
+        if not crash:
             return errno.EINVAL, '', 'crash info: %s not found' % crashid
+        val = json.dumps(crash, indent=4)
         return 0, val, ''
 
     def do_post(self, cmd, inbuf):
         try:
             metadata = self.validate_crash_metadata(inbuf)
         except Exception as e:
             return errno.EINVAL, '', 'malformed crash metadata: %s' % e
-
         crashid = metadata['crash_id']
-        key = 'crash/%s' % crashid
-        # repeated stores of same item are ignored silently
-        if not self.get_store(key):
-            self.set_store(key, inbuf)
+
+        if crashid not in self.crashes:
+            self.crashes[crashid] = metadata
+            key = 'crash/%s' % crashid
+            self.set_store(key, json.dumps(metadata))
+            self._refresh_health_checks()
         return 0, '', ''
 
+    def ls(self):
+        if not self.crashes:
+            self._load_crashes()
+        return self.do_ls({'prefix': 'crash ls'}, '')
+
     def do_ls(self, cmd, inbuf):
-        keys = []
-        for k, meta in self.timestamp_filter(lambda ts: True):
-            entity_name = meta.get('entity_name', 'unknown')
-            keys.append("%s %s" % (k.replace('crash/', ''), entity_name))
-        keys.sort()
-        return 0, '\n'.join(keys), ''
+        if cmd['prefix'] == 'crash ls':
+            r = self.crashes.values()
+        else:
+            r = [crash for crashid, crash in self.crashes.items()
+                 if 'archived' not in crash]
+        if cmd.get('format') == 'json' or cmd.get('format') == 'json-pretty':
+            return 0, json.dumps(r, indent=2), ''
```

Comment:
is there a reason to use `indent=2` here but use `indent=4` in `do_info()`? If not, can this be normalized (I think Ceph uses 4 in most places)